### ISSUE TITLE
Thickness of a stl mesh

### ISSUE THREAD

**Issue by kokokarakolelo** [mikedh/trimesh#1899]
Hello,
I'm a student and I came upon a project with printing 3d files. Before printing the files have to be checked - for which I am using this library. I am trying to calculate the thickness of this 3d model  and check if it is less than 1mm at any point. If it is less than 1mm at any given point then it would be considered unprintable. Do you have any suggestions what can I use?
![image](https://user-images.githubusercontent.com/130614046/231604517-549e205b-e0e9-4176-9009-2578af9a477a.png)
Thanks :).

**Comment by Kiord** [comment]
There is a function `trimesh.proximity.thickness` for that

```python
import trimesh as tm
mesh = tm.load('mesh.stl', process=False)
start_points = mesh.vertices - 0.001 * mesh.vertex_normals # hacky way to get points slightly inside
thickness = tm.proximity.thickness(mesh, start_points, mesh.vertex_normals)

# visualisation with pyvista
import pyvista as pv
pv.wrap(mesh).plot(scalars=thickness)
```

**Comment by kokokarakolelo** [comment]
Hey, thank you for your suggestion! I was wondering how exactly does trimesh.proximity.thickness work? And also how do decide which points are the start points? :)

**Comment by Kiord** [comment]
If you set `method=max_sphere` (default),  it uses `trimesh.proximity.max_tangent_sphere` at each point, the idea is to measure how big can the sphere tangent to the point grow inside the mesh.  It is implemented from [this paper](http://www.cad-journal.net/files/vol_13/CAD_13(2)_2016_199-207.pdf).

If `method=ray`, a ray is cast from each point in the opposite direction of their associated normal and the distance the intersection with the mesh is returned. If it is too slow you can install pyembree (Trimesh will use it under the hood). 

The query points are arbitrary, you can sample points on the surface of the mesh for example, by using `trimesh.sample.sample_surface` or `trimesh.sample.sample_even`.

**Comment by Nosenzor** [comment]
> ```python
> start_points = mesh.vertices - 0.001 * mesh.vertex_normals # hacky way to get points slightly inside
> ```

@Kiord : Is it mandatory to have points slightly inside to measure thickness ?

**Comment by Kiord** [comment]
I tried the function on standford bunny and it led to wrong results without the hack. Perhaps the function should ensure that the points are inside the mesh. This could be a PR I guess.

**Comment by kokokarakolelo** [comment]
@Kiord can I ask you another question? Do you have an idea how can I check if the bottom surface is flat or not - and visualize the results? Or how can I check the surface area of the bottom layer of the insole compared to the surface of the 'top' layer of the insole (presumably where the insole is the widest). One of the things I am struggling to find is if the bottom of the insole is at least 70% of the 'top' surface area of the insole.

**Comment by Kiord** [comment]
`trimesh.curvature.discrete_mean_curvature_measure` is maybe what you are looking for. To select the bottom surface you can use the vertex normals of your well defined mesh.

```python
mesh = ...

up_axis = 1 # y vertical axis
threshold = -0.75
radius = 0.1

insole_mask = mesh.vertex_normals[:, up_axis] < threshold 
mean_curvature = tm.curvature.discrete_mean_curvature_measure(mesh, mesh.vertices, radius)
mean_curvature[~insole_mask] = 0 # we are not interested in the curvature of the upper part

pv.wrap(mesh).plot(scalars=mean_curvature)
```

Else you can use a vertical position threshold instead of an orientation threshold.

**Comment by kokokarakolelo** [comment]
Thank you for the quick reply! 
I tried using `mean_curvature = tm.curvature.discrete_mean_curvature_measure(mesh, mesh.vertices, radius)mean_curvature = tm.curvature.discrete_mean_curvature_measure(mesh, mesh.vertices, radius)` , however doing so I encountered the following: 
![image](https://github.com/mikedh/trimesh/assets/130614046/1e1703db-acd0-4456-bbca-66aa53df3f27)
I tried with `mean_curvature = trimesh.curvature.discrete_gaussian_curvature_measure(mesh, mesh.vertices, radius)
` and it worked just fine! It visualized just what I wanted.
For finding the the position of the insole using the bounding box of the model, however, I could not find if there is a way the change the position of the insole - currently the bottom/top of the insole (the largest side of the bounding box) lays on the z axis and it has to be changed so that it lays on the y axis (for example).
I also tried to check if the insole is watertight and I couldn't really grasp what and how does `mesh.fill_holes()` work. Especially, when `mesh.fill_holes()` first executes and then the other results change - like the visualization of the bottom surface of the insole is no longer the same
